lib/job.js

```diff
@@ -1,106 +1,263 @@
-  expiresAt: 'expires',
-  klassName: 'klass',
-  queueName: 'queue',
-  workerName: 'worker',
-  originalRetries: 'retries',
-  retriesLeft: 'remaining',
-  rawQueueHistory: 'history',
-};
-
-
-// TODO: split up into BaseJob and RecurringJob
+const Promise = require('bluebird').Promise;
+const Path = require('path');
+const logger = require('./logger.js');
+
 class Job {
-  constructor(client, args) {
+
+  constructor(client, config) {
     this.client = client;
-    this._stateChanged = false;
+    this.config = config;
+
+    // Shared between all Job classes
+    this.jid = config.jid;
+    this.priority = config.priority;
+    this.klassName = config.klass;
+    this.queueName = config.queue;
+    this.tags = config.tags;
+    this.data = JSON.parse(config.data);
+
+    // Specific to non-recurring jobs
+    this.state = config.state;
+    this.tracked = config.tracked;
+    this.failure = config.failure;
+    this.history = config.history;
+    this.dependents = config.dependents;
+    this.dependencies = config.dependencies;
+    this.expiresAt = config.expires;
+    this.originalRetries = config.retries;
+    this.retriesLeft = config.remaining;
+    this.workerName = config.worker;
+  }
 
-    // Copy args
-    Object.assign(this, _.mapValues(JOB_ARGS_TO_PROPERTIES, key => args[key]));
+  /**
+   * Given a string name of a class, return the corresponding class
+   *
+   * Split the provided name by `/`, and import the longest working prefix. The
+   * remaining components will be treated as attributes off of the import.
+   *
+   * For example, consider 'some/package/file.js/property/class'. We would
+   * traverse this as:
+   *
+   *  - import 'some/package/file.js/property/class' (failed)
+   *  - import 'some/package/file.js/property', get 'class' attribute (failed)
+   *  - import 'some/package/file.js', get 'property.class' attribute (success)
+   */
+  static import(name) {
+    const segments = name.split(Path.sep);
+    const attributes = [];
+    while (segments.length) {
+      try {
+        const joined = segments.join(Path.sep);
+        logger.debug('Trying to import %s...', joined);
+        // eslint-disable-next-line global-require, import/no-dynamic-require
+        let result = require(joined);
+        attributes.forEach((attribute) => {
+          logger.debug('Getting attribute %s', attribute);
+          result = result[attribute];
+        });
+
+        // If there's no such attribute, break out of the loop and throw
+        if (!result) {
+          break;
+        }
 
-    // Some properties need extra parsing / fixing
-    this.data = JSON.parse(this.data);
-    // Side effect of Lua doing JSON parsing -- Lua empty array is {}
-    if (_.isEmpty(this.tags)) this.tags = [];
-    if (_.isEmpty(this.dependents)) this.dependents = [];
-    if (_.isEmpty(this.dependencies)) this.dependencies = [];
+        logger.debug('Found "%s"', result.name);
+        return result;
+      } catch (err) {
```

Comment:
I think it would be more clear if this `try/catch` were just around the `require` since I believe that's all it's trying to catch.